crates/aptos/src/common/types.rs

```diff
@@ -1683,62 +1739,211 @@ pub struct ArgWithTypeVec {
+    type Error = CliError;
+
+    fn try_into(self) -> Result<Vec<TransactionArgument>, Self::Error> {
+        let mut args = vec![];
+        for arg in self.args {
+            args.push(arg.try_into()?);
+        }
+        Ok(args)
+    }
+}
+
+impl TryInto<Vec<Vec<u8>>> for ArgWithTypeVec {
+    type Error = CliError;
+
+    fn try_into(self) -> Result<Vec<Vec<u8>>, Self::Error> {
+        Ok(self
+            .args
+            .into_iter()
+            .map(|arg_with_type| arg_with_type.arg)
+            .collect())
+    }
+}
+
+impl TryInto<Vec<serde_json::Value>> for ArgWithTypeVec {
+    type Error = CliError;
+
+    fn try_into(self) -> Result<Vec<serde_json::Value>, Self::Error> {
+        let mut args = vec![];
+        for arg in self.args {
+            args.push(arg.to_json()?);
+        }
+        Ok(args)
+    }
+}
+
 /// Common options for constructing an entry function transaction payload.
 #[derive(Debug, Parser)]
 pub struct EntryFunctionArguments {
     /// Function name as `<ADDRESS>::<MODULE_ID>::<FUNCTION_NAME>`
     ///
     /// Example: `0x842ed41fad9640a2ad08fdd7d3e4f7f505319aac7d67e1c0dd6a7cce8732c7e3::message::set_message`
     #[clap(long)]
-    pub function_id: MemberId,
+    pub function_id: Option<MemberId>,
 
+    #[clap(flatten)]
+    pub(crate) type_arg_vec: TypeArgVec,
     #[clap(flatten)]
     pub(crate) arg_vec: ArgWithTypeVec,
 
-    /// TypeTag arguments separated by spaces.
-    ///
-    /// Example: `u8 u16 u32 u64 u128 u256 bool address vector signer`
-    #[clap(long, multiple_values = true)]
-    pub type_args: Vec<MoveType>,
+    /// JSON file specifying public entry function ID, type arguments, and arguments.
+    #[clap(long, parse(from_os_str), conflicts_with_all = &["function-id", "args", "type-args"])]
+    pub(crate) json_file: Option<PathBuf>,
 }
 
 impl EntryFunctionArguments {
-    /// Construct and return an entry function payload from function_id, args, and type_args.
-    pub fn create_entry_function_payload(self) -> CliTypedResult<EntryFunction> {
-        let args: Vec<Vec<u8>> = self
-            .arg_vec
-            .args
-            .into_iter()
-            .map(|arg_with_type| arg_with_type.arg)
-            .collect();
-
-        let mut parsed_type_args: Vec<TypeTag> = Vec::new();
-        // These TypeArgs are used for generics
-        for type_arg in self.type_args.into_iter() {
-            let type_tag = TypeTag::try_from(type_arg.clone())
-                .map_err(|err| CliError::UnableToParse("--type-args", err.to_string()))?;
-            parsed_type_args.push(type_tag)
+    /// Get instance as if all fields passed from command line, parsing JSON input file if needed.
+    fn check_input_style(self) -> CliTypedResult<EntryFunctionArguments> {
+        if self.json_file.is_none() {
+            if self.function_id.is_none() {
+                Err(CliError::CommandArgumentError(
+                    "Must provide either function ID or JSON input file".to_string(),
```

Comment:
Verification at this point is a good safeguard but we should verify this at the clap level with arg group or require + conflicts.

Reply:
https://github.com/aptos-labs/aptos-core/blob/6aa74479f6b3f2b44f06cc0aea7efbd9acff1a51/crates/aptos/src/common/types.rs#L1807